### ISSUE TITLE
System.AccessViolationException in System.Exception.ToString()

### ISSUE THREAD

**Issue by MaGitKo** [dotnet/runtime#59077]
### Description

Sometimes, we get an `AccessViolationException` when formatting an exception as string.
Now it also occurred in an unit test. This is the complete exception stack:

    QueryStudiesAsync_StudyQueryThrowsException_ExceptionIsLogged Aborted: Exit code is -1073741819 (Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
      at System.Diagnostics.StackTrace.GetStackFramesInternal(System.Diagnostics.StackFrameHelper, Int32, Boolean, System.Exception)
      at System.Diagnostics.StackFrameHelper.InitializeSourceInfo(Int32, Boolean, System.Exception)
      at System.Diagnostics.StackTrace.CaptureStackTrace(Int32, Boolean, System.Exception)
      at System.Diagnostics.StackTrace..ctor(System.Exception, Boolean)
      at System.Exception.get_StackTrace()
      at System.Exception.ToString()
      at System.Text.StringBuilder.AppendFormatHelper(System.IFormatProvider, System.String, System.ParamsArray)
      at System.String.FormatHelper(System.IFormatProvider, System.String, System.ParamsArray)
      at System.String.Format(System.String, System.Object, System.Object)
      at syngo.Falcon.Services.Dlos.SyngoShare.SyngoShareStudyProvider+<>c__DisplayClass4_0.<QueryStudiesAsync>b__0()
      at System.Threading.Tasks.Task`1[[System.__Canon, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].InnerInvoke()
      at System.Threading.Tasks.Task+<>c.<.cctor>b__274_0(System.Object)
      at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(System.Threading.Thread, System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
      at System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef, System.Threading.Thread)
      at System.Threading.Tasks.Task.ExecuteEntryUnsafe(System.Threading.Thread)
      at System.Threading.Tasks.Task.ExecuteFromThreadPool(System.Threading.Thread)
      at System.Threading.ThreadPoolWorkQueue.Dispatch()
      at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback())

Occasionally we have this problem also in product code, where the exception stack is identical from the `System.Exception.ToString()` stack frame to the top.

### Configuration

.NET Core 3.1 (tried it with 3.1.17 and 3.1.18)
Windows 10 Enterprise 20H2
x64 Architectue

In our TFS build, the exception does not occur, even though the same .NET Core 3.1 version is used.

Also tried the unit test on my machine with .NET Core 5, here the exception seems not to appear.
But our organization is still on .NET Core 3.1.

### Other information

In issue #10986 I saw that a similar problem was fixed in 2018.

I can provide you with a full crash dump of the unit test process (the "`dotnet test ...`" call).

**Comment by tommcdon** [comment]
@mikem8361

**Comment by austinkinross** [comment]
+1, we're getting something very similar with the .NET 6.0 RC2. This is on ARM64, if it matters.

```
System.Diagnostics.StackTrace.GetStackFramesInternal(System.Diagnostics.StackFrameHelper, Int32, Boolean, System.Exception)
System.Diagnostics.StackFrameHelper.InitializeSourceInfo(Int32, Boolean, System.Exception) [/_/src/coreclr/System.Private.CoreLib/src/System/Diagnostics/StackFrameHelper.cs @ 90]
System.Diagnostics.StackTrace.CaptureStackTrace(Int32, Boolean, System.Exception) [/_/src/coreclr/System.Private.CoreLib/src/System/Diagnostics/StackTrace.CoreCLR.cs @ 62]
System.Diagnostics.StackTrace..ctor(System.Exception, Boolean) [/_/src/libraries/System.Private.CoreLib/src/System/Diagnostics/StackTrace.cs @ 95]
System.Exception.get_StackTrace() [/_/src/libraries/System.Private.CoreLib/src/System/Exception.cs @ 223]
System.Exception.ToString() [/_/src/libraries/System.Private.CoreLib/src/System/Exception.cs @ 128]
[our code that catches the exception]

```

Native stack:

```
coreclr!MethodDesc::GetAssembly+0x24 [src\coreclr\vm\method.hpp @ 1065] 
coreclr!DebugStackTrace::GetStackFramesInternal+0xdf8 [src\coreclr\vm\debugdebugger.cpp @ 482] 
System_Private_CoreLib!System.Diagnostics.StackFrameHelper.InitializeSourceInfo(Int32, Boolean, System.Exception)+0x54 [src/coreclr/System.Private.CoreLib/src/System/Diagnostics/StackFrameHelper.cs @ 92] 
System_Private_CoreLib!System.Diagnostics.StackTrace.CaptureStackTrace(Int32, Boolean, System.Exception)+0x78 [src/coreclr/System.Private.CoreLib/src/System/Diagnostics/StackTrace.CoreCLR.cs @ 62] 
System_Private_CoreLib!System.Diagnostics.StackTrace..ctor(System.Exception, Boolean)+0x2c [src/libraries/System.Private.CoreLib/src/System/Diagnostics/StackTrace.cs @ 96] 
System_Private_CoreLib!System.Exception.get_StackTrace()+0x8c [src/libraries/System.Private.CoreLib/src/System/Exception.cs @ 223] 
System_Private_CoreLib!System.Exception.ToString()+0xd8 [src/libraries/System.Private.CoreLib/src/System/Exception.cs @ 128] 
```

**Comment by mikem8361** [comment]
Does anybody that was reported this problem have a simple repro?  It would be very helpful. Thanks.

**Comment by austinkinross** [comment]
Sorry, if I remember correctly then I didn't have a reliable repro for this.

**Comment by mikem8361** [comment]
Have investigated and have an idea how to fix.  Submitted bug to the C++ group that owns M.D.N.

**Comment by mikem8361** [comment]
This should be fixed in 7.0 now.
